### PR TITLE
Add explanation infidelity

### PR DESCRIPTION
**Pull request by NarineK**
Adding infidelity metric. 
More details about the metric can be found here:  https://arxiv.org/pdf/1901.09392.pdf
Sensitivity will be added in a separate PR.

Since there are many different ways of computing infidelity both for interpretable models and interpreting black-boxes, I decided to name it `infidelity_attr`. This will allow us to add other infidelity metrics in the `infidelity.py` file.

This PR doesn't provide type hints, jit or data parallel tests. Those will be added in separate PRs

In terms of batching: currently the batching is done on the example level. However if the number of perturbed samples increases this might be problematic. That aspect can be addressed in a separate PR.

### CODE REVIEW

**Comment by NarineK** [comment]
#334  needs to get merged first - thus the errors.

**Comment by NarineK** [comment]
Made some structural changes. Will update the PR soon!

File: captum/metrics/_core/infidelity.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+                tensor as input, a single input tensor should be provided.
+                If forward_func takes multiple tensors as input, a tuple
+                of the input tensors should be provided. It is assumed
+                that for all given input tensors, dimension 0 corresponds
+                to the number of examples (aka batch size), and if
+                multiple input tensors are provided, the examples must
+                be aligned appropriately.
+
+        attributions (tensor or tuple of tensors):
+                Attribution scores computed based on an attribution algorithm.
```
Comment (vivekmig) [inline captum/metrics/_core/infidelity.py]: One thing to be careful of with the attributions here is that they should technically be prior to multiplying by (inputs - baselines) for methods that perform this, such as IG and DeepLift. This is because computing the infidelity multiplies the attribution by the perturbation, which essentially assumes that the attribution is expressed in terms of the output change per unit of feature change. This is true for methods such as Saliency, which just take the gradient, but for methods like IG, DeepLift, GradientShap, etc., the attributions should probably be divided by (inputs - baselines) first. 
It would be good to put some explanation / warning to make sure to apply that before passing attributions or even an option which would handle that division internally.

File: tests/metrics/test_infidelity.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    ):
+        def perturb_func(input1, input2=None):
+            perturb1 = torch.stack(
+                [torch.Tensor([0.0009])] * np.prod(list(input1.shape)), dim=-1
+            ).view(input1.shape)
+            if multi_input:
+                perturb2 = torch.stack(
+                    [torch.Tensor([0.0121])] * np.prod(list(input2.shape)), dim=-1
+                ).view(input2.shape)
+                return (perturb1, perturb2), (input1 + perturb1, input2 + perturb2)
```
Comment (vivekmig) [inline tests/metrics/test_infidelity.py]: nit: Based on the signs applied in the infidelity implementation, this maybe should be input1 - perturb1 and input2 - perturb2 or the perturbations should be - perturb1, -perturb2?

File: tests/metrics/test_infidelity.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+            return perturb1, input1 + perturb1
+
+        attrs = attr_algo.attribute(
```
Comment (vivekmig) [inline tests/metrics/test_infidelity.py]: As mentioned above, some attributions may need to be adjusted by inputs - baselines first. Also potentially a good test is to check any method which satisfies completeness, such as IG, and confirm that infidelity for the distribution only containing the baseline (perturbation of input - baseline) should be 0.

File: captum/metrics/_core/infidelity.py (hunk 1 above)
Comment (NarineK) [reply captum/metrics/_core/infidelity.py]: To that point they have a subsection called:
`2.5 Local and Global Explanations`
If it is a local explanation then it shouldn't be multiplied by (input - baseline)
In case of global it still holds if `I` is a subset of inputs - baseline
```
Definition 2.1 are naturally local explanations as I is real-valued. However, this can be
easily modified to a global explanation by multiplying with x - x0 when I is a subset of x - x0.
```

File: captum/metrics/_core/infidelity.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    This specific implementation is primarily tested for attribution-based
+    explanation methods but the idea can be expanded to use for non
+    attribution-based interpretability methods as well.
+
+    Args:
+
+        forward_func (callable):
+                The forward function of our model or any modification of it.
+        perturb_func (callable):
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py]: I actually, thought about it but by allowing to return perturbations and perturbed inputs independently we give users have more flexibility. They can do :
```

perturbed_input = input [operation] perturbation
```
In this case [operation] can be add, subtract, multiply or anything else.

File: captum/metrics/_core/infidelity.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    r"""
+    Explanation infidelity represents the expected mean-squared error
+    between the explanation multiplied by a meaningful input perturbation
+    and the differences between the predictor function at its input
+    and perturbed input.
+    More details about the measure can be found in the following paper:
+    https://arxiv.org/pdf/1901.09392.pdf
+
+    It is derived from the completeness property of well-known attribution
+    algorithms and is an optimal and generalized notion of Sensitivy-n.
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py]: With optimal I meant that it is more efficient to compute than `sens-n`

File: captum/metrics/_core/infidelity.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py]: I think a better example of this is when the perturbation `I` is binary. As explained in the paper. For the global explanations the authors used binary perturbation. In that case `I` is like a mask and in (X - I) is basically the input with ablated features where I is 0.

[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py:57]: I can understand that the users might use addition for example and that can be confusing. In general perturbation can be negative but in that can the negative values would be reflected also in the multiplier `I`.
By taking perturbed input do you mean returning only perturbed input in `perturb_func `?

The problem with only taking perturbed input in case of global is that if the ablation value is not 0 how will we compute the ablation mask ? In that case the user needs to provide `I*Attribution` which would be a bit confusing?

We can potentially document current behavior and also perform asserts in case of local explanation.

[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py:57]: Actually, I guess we can generate the mask based on the differences probably.

File: captum/metrics/_utils/batching.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                        per batch.
+
+    """
+    device = inputs[0].device
+    bsz = inputs[0].size(0)
+
+    if (
+        max_examples_per_batch is not None
+        and max_examples_per_batch // bsz < 1
+        and max_examples_per_batch // bsz > n_samples
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_utils/batching.py:40]: good point! second one should be `or`

File: captum/metrics/_core/infidelity.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+                compute both local and global attributions without a need to
+                perform that division after retrieving the attributions.
+
+                If we want to compute the infidelity of global attributions we
+                can use a binary perturbation matrix that will allow us to select
+                a subset of features from `inputs` or `inputs - baselines` space.
+                This will allow us to approximate sensitivity-n for a global
+                attribution algorithm.
+
+                Attributions will always be max_examples_per_batch
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py:108]: good point! This came though copy paste :D

File: captum/metrics/_core/infidelity.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+                x inputs, with each value
+                providing the attribution of the corresponding input index.
+                If a single tensor is provided as inputs, a single tensor is
+                returned. If a tuple is provided for inputs, a tuple of
+                corresponding sized tensors is returned.
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/metrics/_core/infidelity.py:113]: good point! This came through copy paste :D

File: tests/metrics/test_infidelity.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+def _local_perturb_func(input1, input2=None):
+    perturb1 = torch.stack(
+        [torch.Tensor([0.0009])] * np.prod(list(input1.shape)), dim=-1
+    ).view(input1.shape)
+    if input2 is None:
+        return perturb1, input1 - perturb1
+
+    perturb2 = torch.stack(
+        [torch.Tensor([0.0121])] * np.prod(list(input2.shape)), dim=-1
+    ).view(input2.shape)
```
[the remark this answers is not in the document]
Comment (NarineK) [reply tests/metrics/test_infidelity.py:27]: good point

**Comment by NarineK** [comment]
Addressed comments.
